File: leetcode_contests_train/w464/q3.py
class Solution:
    def maxValue(self, nums: list[int]) -> list[int]:
        n = len(nums)
        dp = [0] * n
        dp[0] = nums[0]
        for i in range(1, n):
            dp[i] = max(nums[i], dp[i - 1])
        min_r = 10**10
        for i in range(n - 1, -1, -1):
            if i + 1 < n and dp[i] > min_r:
                dp[i] = max(dp[i], dp[i + 1])
            min_r = min(min_r, nums[i])
        return dp

File: leetcode_contests_train/w463/q3.go
package main

func minArraySum(nums []int, k int) int64 {
	const MAX = 1 << 32
	dp := make([]int, k+1)
	for i := 1; i <= k; i++ {
		dp[i] = MAX
	}
	res := 0
	for _, num := range nums {
		res += num
		dp[res%k] = min(dp[res%k], res)
		res = dp[res%k]
	}
	return int64(res)
}
